[Infrastructure UI] New Hosts API

crespocarlos opened the pull request:
closes: [#152103](https://github.com/elastic/kibana/issues/152103)
fixes: https://github.com/elastic/kibana/issues/151768

## Summary

This PR adds a new API to return host metrics for the Hosts View page. There are some constraints to what Elasticsearch can do in terms of sorting. The main query runs a `terms` aggregation, whose size is dynamic. 

The caveat of this solution is that instead of sorting by the actual metric aggregation, it sorts by either a `min` or `max` aggregation of a field. The [doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#_ordering_by_a_sub_aggregation) mentions that this is the safest way to sort by a sub-aggregation. This could, however, potentially return a wrong order, but due to Elasticsearch limitations, it's much more likely to return wrong results sorting by an `avg` aggregation.

### Comparison

The comparison below was done with a load of 5 days worth of data, containing 500 hosts on each day, on my local env. For short date ranges, Snapshot API performs consistently better, and the new approach performs better with a large amount of data.

#### Snapshot API

Returns all 500 hosts

15 minutes
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230073865-d0d98ce8-76ff-4295-be9b-94f5b342c9f7.png">

1 day
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230076350-affbad3d-a498-481e-bcba-bcec019ab739.png">

1 week
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230077049-320d90c8-4f3e-47ba-bd16-1c5878a83254.png">

<details>
  <summary>Profiler</summary>

**The profiler results was taken running a query with a 1 day date range.**

<img width="1260" alt="image" src="https://user-images.githubusercontent.com/2767137/230080996-af4d1c21-1118-4c17-9c62-d22e362142d8.png">
<img width="1266" alt="image" src="https://user-images.githubusercontent.com/2767137/230081181-2412025f-d3c7-4dd6-ad19-0637e85fc87f.png">
<img width="1266" alt="image" src="https://user-images.githubusercontent.com/2767137/230081244-000adcbb-1abf-4d09-8116-a5d5b9b19050.png">
<img width="1266" alt="image" src="https://user-images.githubusercontent.com/2767137/230081325-95b9fe6d-1382-4fdd-b138-f31611cbd854.png">
</details>


#### Hosts API

Returns up to 100 hosts

15 minutes
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230073194-cf9b6053-9140-48af-b4b5-0e176285f286.png">

1 day
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230075902-8c1472c0-b007-4f35-a093-bf6bea2d5066.png">

1 week
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230076837-580b2a4a-e444-4a2d-b335-ddd19fa9aec2.png">

<details>
  <summary>Profiler</summary>

**The profiler results was taken running a query with a 1 day date range.**

<img width="1266" alt="image" src="https://user-images.githubusercontent.com/2767137/230080856-29ef575f-a031-4072-86e9-e1e66c03a08a.png">

<img width="1266" alt="image" src="https://user-images.githubusercontent.com/2767137/230080733-e0dcd186-e6de-4884-aede-99cdf7fe9e9a.png">
</details>

### How to test

```bash
curl --location -u elastic:changeme 'http://0.0.0.0:5601/ftw/api/metrics/hosts' \
--header 'kbn-xsrf: xxxx' \
--header 'Content-Type: application/json' \
--data '{
    "limit": 100,
    "metrics": [
        {
            "type": "rx"
        },
        {
            "type": "tx"
        },
        {
            "type": "memory"
        },
        {
            "type": "cpu"
        },
        {
            "type": "diskLatency"
        },
        {
            "type": "memoryTotal"
        }
    ],
    "query": {
        "bool": {
            "must": [],
            "filter": [],
            "should": [],
            "must_not": []
        }
    },
    "timeRange": {
        "from": 1680040800000,
        "to":   1680645600000
    },
    "sortField": "rx",
    "sortDirection": "desc",
    "sourceId": "default"
}'
```

crespocarlos committed:
Create new hosts endpoint

crespocarlos committed:
Query adjustments and refactor

crespocarlos committed:
Refactor new API and use it on the FE

crespocarlos committed:
Integrate API in the Hosts View page

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

crespocarlos committed:
More refactor

crespocarlos committed:
Add tests

crespocarlos committed:
Fix build

crespocarlos committed:
Simplify solution

crespocarlos committed:
Simplify metric objects and bug fix

crespocarlos committed:
Revert FE changes

crespocarlos committed:
Revert FE changes

crespocarlos committed:
Clean up

elasticmachine commented:
Pinging @elastic/infra-monitoring-ui (Team:Infra Monitoring UI)

crespocarlos committed:
More clean ups

jbudz reviewed (approved):
`ftr_configs.yml`

crespocarlos commented on packages/kbn-io-ts-utils/src/route_validation/index.ts:
```diff
@@ -0,0 +1,53 @@
+/*
```
Copied from 
https://github.com/elastic/kibana/blob/main/x-pack/plugins/infra/common/runtime_types.ts#L58

https://github.com/elastic/kibana/blob/main/x-pack/plugins/monitoring/server/lib/create_route_validation_function.ts#L0-L1

The idea is to this function it here and remove those duplications.

crespocarlos committed:
Rename variables

tonyghiani reviewed (changes requested):
First of all, thanks for the effort on this task, I know there is a lot of work and research behind this! 👏

Reading the implementation and playing with the endpoint I found some things behaving differently from how we would expect on the client side and I've noticed some areas of improvements:
- Changing `sourceId` to something meaningless doesn’t break the request and silently returns an empty array. Shouldn’t it be a `400 Bad request` informing about a missing source id? Should the sourceId be a path param instead of a body param?
- Timerange only accepts timestamps? Can’t we make it flexible to accept also dates and relative formats as other APIs do? It could be easier for us to deal with it at the moment of passing relative dates.
- In metadata `host.os.name` is null, but `cloud.provider` is an empty string. Shouldn’t be consistent and return null as well if missing?
- I had some first issues using the API before reading the code, I haven't clear what is the format type of each returned metric and what params I could use. Would be awesome if you could add documentation for the endpoint and some usage examples.
- Can’t sort by `host.os.name`? This is currently a feature we have in the client, would it break if we can't sort by this field?
- Looks like the filtering is not working with some hosts. I tried on a lite-edge cluster to filter by one of the hosts that are currently in the table (`opbeans-php-d994b56-5ndtq`) but it returned an empty array as if this hostname didn't exist. For other hosts it worked correctly.
```
POST kbn:/api/metrics/hosts
{
  "limit": 100,
  "metrics": [
    {
      "type": "rx"
    },
    {
      "type": "tx"
    },
    {
      "type": "memory"
    },
    {
      "type": "cpu"
    },
    {
      "type": "diskLatency"
    },
    {
      "type": "memoryTotal"
    }
  ],
  "query": {
    "bool": {
      "must": [],
      "filter": [
        {
          "terms": {
            "host.name": [
              "opbeans-php-d994b56-5ndtq"
            ]
          }
        }
      ],
      "should": [],
      "must_not": []
    }
  },
  "timeRange": {
    "from": 1680762352422,
    "to": 1680765952422
  },
  "sortField": "name",
  "sortDirection": "asc",
  "sourceId": "default"
}
```
- I’m wondering if we could sort by multiple fields if necessary. Having an API similar to the ES one, we could potentially sort the results by multiple fields easily. Something like a param `sort: [{rx: “desc”}, {tx: “asc”}]`
- Removing the query body option fails the request. Shouldn’t be optional and return all the limited results if no query is applied?

I made a first pass on the PR and I'll spend some more time reviewing it, let me know if you have any doubt about my thoughts :)

crespocarlos commented on x-pack/plugins/infra/common/http_api/hosts/get_hosts.ts:
```diff
@@ -0,0 +1,91 @@
+
+const LimitRangeRT = rt.brand(
+  rt.number, // codec
+  (n): n is LimitRange => n > 0 && n <= 100,
```
0 is not accepted. It breaks the terms aggregation

crespocarlos commented:
Hey @tonyghiani ! I'll answer these questions before addressing your comments.

>Changing sourceId to something meaningless doesn’t break the request and silently returns an empty array. Shouldn’t it be a 400 Bad request informing about a missing source id? Should the sourceId be a path param instead of a body param?

Hm, it'd be confusing to have this as path params, but it's fair to say that returning a 400 if sourceId Is invalid.

>Timerange only accepts timestamps? Can’t we make it flexible to accept also dates and relative formats as other APIs do? It could be easier for us to deal with it at the moment of passing relative dates.

I rather not give so much flexibility here. If the client passes relative dates e.g: `gte: now-15m, lte: now`, elasticsearch won't be able to cache the results, because `now` will change each time the query is executed (see [this](https://github.com/elastic/elasticsearch/issues/4947)). Caching will be particularly importing if you sort by any column, we want elasticsearch to cache results as much as possible.

> In metadata host.os.name is null, but cloud.provider is an empty string. Shouldn’t be consistent and return null as well if missing?

yeah. makes sense

>I had some first issues using the API before reading the code, I haven't clear what is the format type of each returned metric and what params I could use. Would be awesome if you could add documentation for the endpoint and some usage examples.

> Can’t sort by host.os.name? This is currently a feature we have in the client, would it break if we can't sort by this field?

Unfortunately, it's not possible because it's a top_metric aggregation. This column will have to be removed from the table, or at least not let the user sort by it.

> Looks like the filtering is not working with some hosts. I tried on a lite-edge cluster to filter by one of the hosts that are currently in the table (opbeans-php-d994b56-5ndtq) but it returned an empty array as if this hostname didn't exist. For other hosts it worked correctly.

I'll try to reproduce the issue.

crespocarlos commented on x-pack/plugins/infra/common/http_api/hosts/get_hosts.ts:
```diff
@@ -0,0 +1,91 @@
+
+export const HostSortFieldRT = rt.keyof({ name: null, ...HostMetricTypeRT.keys });
+
+interface LimitRangeBrand {
+  readonly LimitRange: unique symbol;
+}
+
+export type LimitRange = rt.Branded<number, LimitRangeBrand>;
+
+const LimitRangeRT = rt.brand(
+  rt.number, // codec
+  (n): n is LimitRange => n > 0 && n <= 100,
+  // refinement of the number type
+  'LimitRange' // name of this codec
+);
```
Yeah. Good idea!

crespocarlos commented on x-pack/plugins/infra/server/routes/hosts/lib/utils.ts:
```diff
@@ -0,0 +1,51 @@
+  );
+};
+
+export const parseFilters = (query: any): estypes.QueryDslQueryContainer => {
+  const parsed = isValidFilter(query) ? query : undefined;
+  if (!parsed) {
+    throw Boom.badRequest('Invalid query');
+  }
+
+  return parsed;
+};
```
It was originally a parser. I forgot to rename it. Thanks for the suggestion

tonyghiani commented:
Hey @crespocarlos thanks for clarifying all the previous questions!
Just one more doubt regarding
> I rather not give so much flexibility here. If the client passes relative dates e.g: gte: now-15m, lte: now, elasticsearch won't be able to cache the results, because now will change each time the query is executed (see https://github.com/elastic/elasticsearch/issues/4947). Caching will be particularly importing if you sort by any column, we want elasticsearch to cache results as much as possible.

When the timerange is relative (`gte: now-15m, lte: now`) when still set from the client timestamps differents each time the request is performed, so the query won't be cached anyway right? I guess what elasticsearch caches is absolute dates, while for the relative ones won't be able to do it, but this should be reflected anyway on how we pass the time range from the client. However, it was mostly for having more flexibility on this front, not a blocking feature request for me 👌

crespocarlos commented on x-pack/plugins/infra/server/routes/hosts/lib/utils.ts:
```diff
@@ -0,0 +1,51 @@
+  const parsedFilters = parseFilters(query);
+
+  // ignores minimum_should_match
+  return Object.entries(parsedFilters.bool ?? {})
```
Not really though, because the validation checks if the object is compliant with the `QueryDslQueryContainer`, which defines `bool` as optional.

crespocarlos replied:
Solved if I create new type

crespocarlos commented on x-pack/plugins/infra/server/routes/hosts/lib/utils.ts:
```diff
@@ -0,0 +1,51 @@
+  if (!boolClause) {
+    return false;
+  }
+  return (
+    boolClause.filter !== undefined ||
+    boolClause.must !== undefined ||
+    boolClause.must_not !== undefined ||
+    boolClause.should !== undefined
+  );
```
you meant `some` instead of `every` here, right?

tonyghiani replied:
I meant `every`, cause any of the 4 available fields should either be an array or undefined, any different value is not accepted.
```js
// ✅
{
  filter: [],
  must: [],
  must_not: [],
  should: []
}

// ✅
{
  filter: [],
  must: []
}

// ❌
{
  filter: {},
  must: [],
  must_not: [],
  should: []
}

// ❌
{
  filter: {},
  must: []
}
```

crespocarlos replied:
As long as one of the is present, the code should consider it a valid object. This should be a valid filter:
```ts
{
  filter: [{
    terms: {
       'host.name': 'host-0'
    }],
}
```

Besides, the type definition of `filter`, `must`, `must_not`, and `should` is `QueryDslQueryContainer | QueryDslQueryContainer[]`. We could say in this API that we only accept `QueryDslQueryContainer[]` and be as strict that they must be arrays

crespocarlos commented:
> When the timerange is relative (gte: now-15m, lte: now) when still set from the client timestamps differents each time the request is performed

Indeed, but if you sort by a column, we wouldn't necessarily pass a new timestamp. We would use what's stored in the state and just pass the `sortField` and `sortDirection`

crespocarlos commented:
> Looks like the filtering is not working with some hosts. I tried on a lite-edge cluster to filter by one of the hosts that are currently in the table (opbeans-php-d994b56-5ndtq) but it returned an empty array as if this hostname didn't exist. For other hosts it worked correctly.

The filter was filtering too much. There was a clause to filter `event.module: "system"`

crespocarlos commented:
> Changing sourceId to something meaningless doesn’t break the request and silently returns an empty array. Shouldn’t it be a 400 Bad request informing about a missing source id? Should the sourceId be a path param instead of a body param?

Passing a random `sourceId` is not invalid per se. If InfraSources has trouble returning a source configuration based on the `sourceId`, it in the end returns a fallback object, which might work in some cases. I'd probably not return a 404 in this case.

tonyghiani commented:
> Passing a random `sourceId` is not invalid per se. If InfraSources has trouble returning a source configuration based on the `sourceId`, it in the end returns a fallback object, which might work in some cases. I'd probably not return a 404 in this case.

This is why I also thought about passing the `sourceId` in the path. Similar to how the `GET /api/metrics/source/${sourceId}` works, when a passed `sourceId` doesn't match any source it returns a 404 because the resources don't exist.
It is pretty unopinionated, I do see this endpoint as "given a sourceId, retrieve hosts metrics for the related source", which is why I'm thinking to pass it as a param and return 404 in case there is no match, but we could use some other opinion by the team.

crespocarlos committed:
CR fixes

crespocarlos commented:
>  This is why I also thought about passing the sourceId in the path. Similar to how the GET /api/metrics/source/${sourceId} works, when a passed sourceId doesn't match any source it returns a 404 because the resources don't exist.
It is pretty unopinionated, I do see this endpoint as "given a sourceId, retrieve hosts metrics for the related source", which is why I'm thinking to pass it as a param and return 404 in case there is no match, but we could use some other opinion by the team.

I'm not really familiar with this endpoint, but isn't it supposed to return *the* configuration based on the sourceId - meaning that we're looking at the `source` collection and returning it by an id?

In this case, I wouldn't say hosts is a resource from `source` collection for it to be in a path param. Of course, this is open to interpretation, but IMO it would also mean that every other endpoint that receives a `sourceId` in infra would be part of the `source` collection.

crespocarlos committed:
in_range_rt import fix

tonyghiani commented:
> I'm not really familiar with this endpoint, but isn't it supposed to return _the_ configuration based on the sourceId - meaning that we're looking at the `source` collection and returning it by an id?
> 
> In this case, I wouldn't say hosts is a resource from `source` collection for it to be in a path param. Of course, this is open to interpretation, but IMO it would also mean that every other endpoint that receives a `sourceId` in infra would be part of the `source` collection.
> 
> You could on the other hand hypothetically have something like `/api/metrics/hosts/host-0` - which seems to be the same idea in the `/metrics/source` endpoint.

Yep agree, I guess what concern me is not explicitly seeing why I'm not getting any metric if the `sourceId` is misspelt or missing

crespocarlos committed:
revert in_range_rt in kbn-io-ts-utils

crespocarlos commented on x-pack/plugins/infra/server/routes/hosts/README.md:
```diff
@@ -0,0 +1,92 @@
+# Infra Hosts API
```
I wish we had open-api docs in place

crespocarlos committed:
Move InRangeRT to kbn-io-ts-utils

crespocarlos committed:
Adjustments in source configuration usage and added more tests

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 152103-new-hosts-api

crespocarlos committed:
Fix README.md

crespocarlos committed:
Remove filter by event.module: system

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 152103-new-hosts-api

crespocarlos commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 152103-new-hosts-api

crespocarlos commented:
Closing this PR to rethink the solution
